cmd/gpu_plugin/fakedev/generator.go

```diff
@@ -0,0 +1,298 @@
+	}
+	opts.dirs++
+
+	data := []byte(fmt.Sprintf("%d", opts.DevMemSize))
+	file := fmt.Sprintf("%s/lmem_total_bytes", base)
+
+	if err := os.WriteFile(file, data, fileMode); err != nil {
+		return err
+	}
+	opts.files++
+
+	path := fmt.Sprintf("%s/device/drm/card%d", base, card)
+	if err := os.MkdirAll(path, dirMode); err != nil {
+		return err
+	}
+	opts.dirs++
+
+	path = fmt.Sprintf("%s/device/drm/renderD%d", base, renderBase+i)
+	if err := os.Mkdir(path, dirMode); err != nil {
+		return err
+	}
+	opts.dirs++
+
+	data = []byte("0x8086")
+	file = fmt.Sprintf("%s/device/vendor", base)
+
+	if err := os.WriteFile(file, data, fileMode); err != nil {
+		return err
+	}
+	opts.files++
+
+	node := 0
+	if opts.DevsPerNode > 0 {
+		node = i / opts.DevsPerNode
+	}
+
+	data = []byte(fmt.Sprintf("%d", node))
+	file = fmt.Sprintf("%s/device/numa_node", base)
+
+	if err := os.WriteFile(file, data, fileMode); err != nil {
+		return err
+	}
+	opts.files++
+
+	if opts.VfsPerPf > 0 && i%(opts.VfsPerPf+1) == 0 {
+		data = []byte(fmt.Sprintf("%d", opts.VfsPerPf))
+		file = fmt.Sprintf("%s/device/sriov_numvfs", base)
+
+		if err := os.WriteFile(file, data, fileMode); err != nil {
+			return err
+		}
+		opts.files++
+	}
+
+	for tile := 0; tile < opts.TilesPerDev; tile++ {
+		path := fmt.Sprintf("%s/gt/gt%d", base, tile)
+		if err := os.MkdirAll(path, dirMode); err != nil {
+			return err
+		}
+		opts.dirs++
+	}
+
+	return nil
+}
+
+func addDevfsDriTree(root string, opts *genOptions, i int) error {
+	base := fmt.Sprintf("%s/dri", root)
+	if err := os.MkdirAll(base, dirMode); err != nil {
+		return err
+	}
+	opts.dirs++
+
+	mode := uint32(fileMode | devNullType)
+	devid := int(unix.Mkdev(uint32(devNullMajor), uint32(devNullMinor)))
+
+	file := fmt.Sprintf("%s/card%d", base, cardBase+i)
+	if err := unix.Mknod(file, mode, devid); err != nil {
+		return err
+	}
+	opts.devs++
+
+	file = fmt.Sprintf("%s/renderD%d", base, renderBase+i)
+	if err := unix.Mknod(file, mode, devid); err != nil {
+		return err
+	}
+	opts.devs++
+
+	return nil
+}
+
+func addDebugfsDriTree(root string, opts *genOptions, i int) error {
+	base := fmt.Sprintf("%s/kernel/debug/dri/%d", root, i)
+	if err := os.MkdirAll(base, dirMode); err != nil {
+		return err
+	}
+	opts.dirs++
+
+	path := fmt.Sprintf("%s/i915_capabilities", base)
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, fileMode)
+
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	opts.files++
+
+	// keys are in random order which provides extra testing for NFD label parsing code
+	for key, value := range opts.Capabilities {
+		line := fmt.Sprintf("%s: %s\n", key, value)
+		if _, err = f.WriteString(line); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// generateDriFiles generates the fake sysfs + debugfs + devfs dirs & files according to given options.
+func generateDriFiles(opts genOptions) {
+	if opts.Info != "" {
+		log.Printf("Config: '%s'", opts.Info)
+	}
+
+	entries, _ := os.ReadDir(devfsPath)
```

Comment:
shouldn't there be error handling for the ReadDir -calls in this function?

Reply:
"If an error occurs reading the directory, ReadDir returns the entries it was able to read before the error, along with the error."

If there are any entries, they are removed few lines later [1].  Directory missing does not matter, as there's MkdirDirAll() later on for the same location, and that does have a check.

Only possible issue I can see is there being items, reading failing on first item, so that none of them are reported nor removed, and new config being different than earlier one.  Later activity would then most likely fail too, but I can add error check already for ReadDir.

[1] Removal probably needs a check too, as config could be changed to something with less devices, and device creation (following the removal of pre-existing files) would not necessarily fail if only extra devices were write-protected => I'll check removals too.

There's still the issue of something mangling content of the fake devfs dir in parallel to the pod, as container runtime requires it to be mounted from host (whereas fake sysfs can be pod-internal emptyDir).  That can be fixed only on Kubernetes side, by some (new) feature providing a node-local volume that Pods can share securely.